Handling events after transaction completion using a separate session

pedroigor opened the pull request:
Closes #15656

* Sending e-mails using a separate session after completion in order to avoid fetching data from the database from an already closed session

As discussed in #15656, ideally we should send e-mails asynchronously but that would require reviewing not only this provider but any other enlisting work to be executed after transaction completion. It also needs more discussion around the design and how to properly support async tasks after completion.

pedroigor committed:
Handling events after transaction completion using a separate session

Closes #15656

pedroigor commented:
@sventorben Could you try it out, if you are able to build from sources?

mposolda reviewed (dismissed):
@pedroigor LGTM, Thanks!

BTV. for simulating the issue, maybe you can also possibly try to test with cache disabled if needed?

svenseeberg commented:
> @sventorben Could you try it out, if you are able to build from sources?

I guess you're referring to me? Yes, we will.

pedroigor commented:
> I guess you're referring to me? Yes, we will.

Oops :)

pedroigor commented:
> @pedroigor LGTM, Thanks!
> 
> BTV. for simulating the issue, maybe you can also possibly try to test with cache disabled if needed?

It is non-deterministic. It depends on the user/realm cached state. I think it might either bring instability or not properly assert the behavior.

svenseeberg commented:
We applied the patch of this PR to the 20.0.3 tag to avoid any migrations. We're currently running this in our production environment to see if the error occurs. The initial error (https://github.com/keycloak/keycloak/issues/15656) is gone.

We noticed a new error that did not yet occur in our logs, which we think is related to this PR. However, it could also be due to our naive approach of applying the patch to the 20.0.3 release.

```
2023-02-09 13:39:55,376 ERROR [org.keycloak.events.email.EmailEventListenerProvider] (executor-thread-1) Failed to send type mail: org.keycloak.email.EmailException: Failed to template email
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.processTemplate(FreeMarkerEmailTemplateProvider.java:246)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.send(FreeMarkerEmailTemplateProvider.java:261)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.send(FreeMarkerEmailTemplateProvider.java:256)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.send(FreeMarkerEmailTemplateProvider.java:207)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.sendEvent(FreeMarkerEmailTemplateProvider.java:110)
	at org.keycloak.events.email.EmailEventListenerProvider$1.run(EmailEventListenerProvider.java:79)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:260)
	at org.keycloak.events.email.EmailEventListenerProvider.sendEmail(EmailEventListenerProvider.java:71)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.keycloak.events.EventListenerTransaction.commitImpl(EventListenerTransaction.java:62)
	at org.keycloak.models.AbstractKeycloakTransaction.commit(AbstractKeycloakTransaction.java:48)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:146)
	at org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler.close(TransactionalSessionHandler.java:64)
	at org.keycloak.quarkus.runtime.integration.jaxrs.TransactionalResponseFilter.filter(TransactionalResponseFilter.java:47)
	at org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl.filter(ContainerResponseContextImpl.java:367)
	at org.jboss.resteasy.core.ServerResponseWriter.executeFilters(ServerResponseWriter.java:252)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:101)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:74)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeResponse(SynchronousDispatcher.java:594)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:524)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at org.keycloak.theme.DefaultThemeSelectorProvider.getThemeName(DefaultThemeSelectorProvider.java:40)
	at org.keycloak.theme.DefaultThemeManager.getTheme(DefaultThemeManager.java:60)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.getTheme(FreeMarkerEmailTemplateProvider.java:251)
	at org.keycloak.email.freemarker.FreeMarkerEmailTemplateProvider.processTemplate(FreeMarkerEmailTemplateProvider.java:212)
	... 49 more
```

pedroigor commented:
@svenseeberg Looks unrelated. Is this error happening when starting or at runtime? I guess runtime, right?

pedroigor commented:
@svenseeberg It is not, my fault. Will send a fix.

Our test suite is supposed to catch the behavior, and will also look at why it did not do it.

pedroigor commented:
@svenseeberg It should be fine now.

pedroigor commented:
@mposolda If you can review it again, we were missing a test to cover the email listener provider. Added one. That is why we did not catch the error herein reported.

I just added the code to initialize the context with the realm, optionally the client, and also make the request available when calling downstream providers.
